### ISSUE TITLE
How to integrate rhd-frontend assets with Drupal

### ISSUE THREAD

**Issue by jordanpagewhite** [redhat-developer/rhd-frontend#127]
Drupal is the primary (only?) customer of the `rhd-frontend` project and needs to integrate and consume the assets produced by the `rhd-frontend` build process. The `rhd-frontend` project should aim to produce assets in a way that are easy for Drupal to consume and integrate with its [libraries style of JS and CSS asset management](https://www.drupal.org/docs/8/theming/adding-stylesheets-css-and-javascript-js-to-a-drupal-8-theme) for the new theme.

Idealised, `rhd-frontend` will produce assets in the following structure that Drupal can consume:

* `rhd.global.min.js`
   * All scripts that need to be loaded on every 'page'/in all contexts
* `rhd.global.min.css`
  * All styles that need to be loaded on every 'page'/in all contexts
* `rhd.[component].min.js`
   * Scripts required by this component. If `[component]` doesn't make clear the context(s) in which these scripts need to be loaded, please document those contexts.
* `rhd.[component].min.css`
  * Styles required by this component. If `[component]` doesn't make clear the context(s) in which these styles need to be loaded, please document those contexts.

For example, we will likely/should have something like `rhd.card.min.js` and `rhd.card.min.css`. With these card-specific styles, we can create a `card` library in Drupal and attach that library if and only if a component that uses the Card component is loaded.

### Acceptance Criteria

* Agree and document the structure in which `rhd-frontend` will produce assets to be consumed by Drupal
* Update the `rhd-frontend` to build out the assets in a form that can be consumed by Drupal.
* Determine the best method of sharing the assets from the `rhd-frontend` build with Drupal.

**Comment by robpblake** [comment]
@KamiQuasi Once we've agreed on the format in which we'll produce the assets from the `rhd-frontend` component, I can start to design out the CDN space if we still think we can use it.

cheers,

Rob

**Comment by KamiQuasi** [comment]
Drupal is the primary consumer of the front-end files, but currently Wordpress and Keycloak also utilize copies of the front-end styles and scripts to mimic the look and feel of the site. Obviously, those additional consumers may stop at some point, but they have similar needs, so I don't know that it changes any of the planning for consumables.

I like the naming convention you've suggested as well, @jordanpagewhite, how would it be best for people to communicate any dependencies that would be needed per component?

**Comment by KamiQuasi** [comment]
Re: Folder structure and paths
I believe stand-alone files will exist within sub-folders within the main directories as well. So you might have `css-min/components/rhd.card.min.css` and `js/card/card.min.js` or some such thing. Regardless of that particular structure, the `dist` directory is where any production-level assets would be built. Does that help @robpblake and @jordanpagewhite ?

**Comment by robpblake** [comment]
@KamiQuasi Yes that definitely helps. If we can reach agreement on the build for `rhd-frontend` building out the component level JS & CSS and then also a global CSS & JS file that are minimised, then I think that gives @jordanpagewhite the flexibility he needs to only include required JS/CSS on certain Drupal pages.

WRT the CDN side of this, I'll simply make the guarantee that all component JS & CSS and the global CSS & JS will be on the CDN in the directory of the release. So lets say we cut release `1234`, I'll guarantee that you'll be able to load:

- https://prod.redhatstatic.com/developers/fe/1234/global/global.min.js
- https://prod.redhatstatic.com/developers/fe/1234/global/global.min.css
- https://prod.redhatstatic.com/developers/fe/1234/component/foo/foo.min.css
- https://prod.redhatstatic.com/developers/fe/1234/component/foo/foo.min.js

Are we comfortable with that? (cc @jordanpagewhite)

cheers,

Rob

**Comment by KamiQuasi** [comment]
@robpblake currently the folder structure will end up being by filetype with `dist/js` and `dist/css` folders unless we restructure the src directory to be by component, which we could definitely do. @jordanpagewhite would restructuring by component be helpful? I definitely like the idea of more self-contained components.

**Comment by jordanpagewhite** [comment]
Cool, yes, that sounds good to me @robpblake . @KamiQuasi and @robpblake I don't particularly care about the directory per component. It seems nice to me, but I don't thinks it actually impacts whether or not we _can_ do it. Does it make it easier for either of you? If so, I am cool with that.

**Comment by robpblake** [comment]
@jordanpagewhite @KamiQuasi Doesn't really make any difference to me. Only thing I care about is that we're all agreed on the structure and can go from there.

Once we're agreed on what's the easiest thing to do, I'll ensure that we match that on the CDN.

cheers,

Rob

**Comment by jordanpagewhite** [comment]
@KamiQuasi and @robpblake I am not sure that it would be wise to invest the time to component-ize our assets and build processes in this way at this time. Given our impending deadline, I think that we should likely just focus on hitting the deadline first and then if we have time, we can of course return and confidently component-ize these dist files within the deadline. Otherwise, we should likely make the effort to component-ize the dist files after September.

**Comment by jordanpagewhite** [comment]
This is a worthwhile conversation to continue, but cannot continue to be in the scope of this DX Surge.
